### PR TITLE
Enable FusedSDPA for Mpt

### PR DESCRIPTION
**Pull request by Jianhong-Zhang**
Enable FusedSDPA for Mpt

Signed-off-by: Jianhong Zhang <jianhong.zhang@intel.com>
co-authored-by: Ling, Jing1 <jing1.ling@intel.com>

### CODE REVIEW

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 1]
```diff
@@ -90,13 +126,38 @@ def gaudi_mpt_attention_forward(
...
+        import habana_frameworks.torch.hpu as ht
+
+        with ht.sdp_kernel(enable_recompute=flash_attention_recompute):
+            if query_length > 8192:
+                attn_output = self.gaudi_flash_attn_v1(
+                    query_states, key_states, value_states, attention_mask, 0.0, self.block_size
+                )
+                htcore.mark_step()
+            else:
+                attn_output = scaled_dot_product_attention(
```
Comment (schoi-habana) [inline optimum/habana/transformers/models/mpt/modeling_mpt.py:139]: this function is a naive implementation of sdpa. you need fused sdpa instead

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 2]
```diff
@@ -90,13 +126,38 @@ def gaudi_mpt_attention_forward(
...
 
-    # (batch_size, n_heads, seq_length, key_length)
-    attn_weights = nn.functional.softmax(attention_scores.float(), dim=-1).to(value_states.dtype)
-    attn_weights = nn.functional.dropout(attn_weights, p=self.attn_dropout_p, training=self.training)
+    if use_flash_attention and FusedSDPA:
+        import habana_frameworks.torch.hpu as ht
+
+        with ht.sdp_kernel(enable_recompute=flash_attention_recompute):
+            if query_length > 8192:
+                attn_output = self.gaudi_flash_attn_v1(
```
Comment (schoi-habana) [inline optimum/habana/transformers/models/mpt/modeling_mpt.py:134]: where is self.gaudi_flash_attn_v1 defined?

**Review by schoi-habana — CHANGES_REQUESTED** [review]

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 2 above)
Comment (Jianhong-Zhang) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py]: I followed qwen2, and didn't find in any other models. Looks it is not needed in this model, so I removed it

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 3]
```diff
@@ -90,13 +101,32 @@ def gaudi_mpt_attention_forward(
...
         attention_scores = attention_scores.masked_fill(attention_mask, torch.finfo(query_states.dtype).min)
 
-    # (batch_size, n_heads, seq_length, key_length)
-    attn_weights = nn.functional.softmax(attention_scores.float(), dim=-1).to(value_states.dtype)
-    attn_weights = nn.functional.dropout(attn_weights, p=self.attn_dropout_p, training=self.training)
+    if use_flash_attention and FusedSDPA:
+        import habana_frameworks.torch.hpu as ht
+
+        with ht.sdp_kernel(enable_recompute=flash_attention_recompute):
+            attn_output = FusedSDPA.apply(
```
Comment (schoi-habana) [inline optimum/habana/transformers/models/mpt/modeling_mpt.py:108]: fusedSDPA should replace the whole attention block when --use_flash_attention is enabled.
but this commit inserted fusedSDPA in the middle of attention after q*k.
you should be very careful when you make changes in the model and run thorough tests

**Review by schoi-habana — CHANGES_REQUESTED** [review]

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 3 above)
Comment (Jianhong-Zhang) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:101]: updated

**Comment by Jianhong-Zhang** [comment]
No more long graph compilation issue observed with newer docker version for FusedSDPA.

  docker_version: Docker version : 1.17.0
  pt_version: PyTorch version : 2.3.1
  build_version: Build version :488

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 4]
```diff
@@ -26,23 +26,34 @@
...
+
+import habana_frameworks.torch.core as htcore
 
 def gaudi_mpt_attention_forward(
     self,
     hidden_states: torch.Tensor,
     position_bias: torch.Tensor,
     past_key_value: Optional[Tuple[torch.Tensor]] = None,
     attention_mask: Optional[torch.Tensor] = None,
+    output_attentions: bool = False,
```
Comment (schoi-habana) [inline optimum/habana/transformers/models/mpt/modeling_mpt.py:45]: output_attentions is never passed from mpt_block_forward(). do you need to add it?

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 5]
```diff
@@ -85,18 +94,39 @@ def gaudi_mpt_attention_forward(
...
-    # (batch_size, n_heads, seq_length, key_length)
-    attn_weights = nn.functional.softmax(attention_scores.float(), dim=-1).to(value_states.dtype)
-    attn_weights = nn.functional.dropout(attn_weights, p=self.attn_dropout_p, training=self.training)
+    attn_output = attn_output.permute(0, 2, 1, 3).contiguous().view(batch_size, seq_length, -1)
+    attn_output = self.out_proj(attn_output)
 
-    context_states = torch.matmul(attn_weights, value_states)
-    context_states = context_states.permute(0, 2, 1, 3).contiguous().view(batch_size, seq_length, -1)
-    attn_output = self.out_proj(context_states)
+    if not output_attentions:
```
Comment (schoi-habana) [inline optimum/habana/transformers/models/mpt/modeling_mpt.py:128]: seems redundant

**Review by schoi-habana — CHANGES_REQUESTED** [review]

**Review by schoi-habana — APPROVED** [review]
lgtm

**Comment by atakaha** [comment]
Could you provide command line and some data with/without FusedSDPA?

**Comment by Jianhong-Zhang** [comment]
> Could you provide command line and some data with/without FusedSDPA?

Details refer to https://habana.atlassian.net/browse/GS-329
Any other question, we can discuss offline

**Review by mandy-li — APPROVED** [review]

**Comment by mandy-li** [comment]
@regisss , please review and merge if ok for you. thanks

**Comment by Jianhong-Zhang** [comment]
@regisss, can you please review the pr? thanks

**Review by regisss — APPROVED** [review]

**Pull request merged**